osutil/env.go

```diff
@@ -50,3 +53,56 @@ func GetenvBool(key string, dflt ...bool) bool {
 
 	return b
 }
+
+// SubstituteEnv takes a list of environment strings like:
+// - K1=BAR
+// - K2=$BAR
+// - K3=${BAZ}
+// and substitutes them using the os environment strings.
+//
+// Strings that do not have the form "k=v" will be dropped.
+//
+// The result will be sorted list of environment strings.
+//
+// Circular references (A=$B,B=$A) will result in empty A,B
+// (just like in shell)
+func SubstituteEnv(env []string) []string {
+	envMap := map[string]string{}
+	for _, s := range env {
+		l := strings.SplitN(s, "=", 2)
+		if len(l) < 2 {
+			continue
+		}
+		envMap[l[0]] = l[1]
+	}
+
+	// this always terminates, in each iteration of the loop we
+	// eliminate at least one value with a "$var"
```

Comment:
Thank you! I updated the code and it makes heavy use of the spread environment code now. I called it `strutil.OrderedMap` but I'm happy to adjust as needed. There are some tiny derivations from the spread code but happy to fix/update if I missed something (or if you disagree with the derivations).